rbpf-cli/src/coverage.rs

```diff
@@ -0,0 +1,252 @@
+                    let section_bytes = &elf_bytes[file_range.start..file_range.end];
+                    debug!("Section {}: {} bytes", id.name(), section_bytes.len());
+                    Cow::Borrowed(section_bytes)
+                }
+            })
+        };
+
+        // Teach gimli how to switch endianness when needed.
+        let borrow_section: &dyn for<'a> Fn(
+            &'a Cow<[u8]>,
+        )
+            -> gimli::EndianSlice<'a, gimli::RunTimeEndian> =
+            &|section| gimli::EndianSlice::new(&*section, gimli::RunTimeEndian::Little);
+
+        // Load all of the sections.
+        let dwarf_cow = Dwarf::load(&load_section)?;
+
+        // Create `EndianSlice`s for all of the sections.
+        let dwarf = dwarf_cow.borrow(&borrow_section);
+
+        let mut cov = Self::default();
+
+        // Iterate over the compilation units.
+        let mut iter = dwarf.units();
+        while let Some(header) = iter.next()? {
+            if let Err(e) = cov.process_unit(&dwarf, header, text_range.start as u64, &hits) {
+                error!("Failed to extract coverage from compile unit: {:?}", e);
+            }
+        }
+
+        Ok(cov)
+    }
+
+    fn process_unit(
+        &mut self,
+        dwarf: &Dwarf<EndianSlice<'_, RunTimeEndian>>,
+        header: UnitHeader<EndianSlice<'_, RunTimeEndian>, usize>,
+        text_section_offset: u64,
+        hits: &bv::BitVec,
+    ) -> Result<(), Box<dyn std::error::Error>> {
+        debug!(
+            "Line number info for unit at <.debug_info+0x{:x}>",
+            header.offset().as_debug_info_offset().unwrap().0
+        );
+        let unit = dwarf.unit(header)?;
+
+        // Get the line program for the compilation unit.
+        let program = match unit.line_program.clone() {
+            None => return Ok(()),
+            Some(program) => program,
+        };
+
+        let comp_dir = if let Some(ref dir) = unit.comp_dir {
+            PathBuf::from(dir.to_string_lossy().into_owned())
+        } else {
+            PathBuf::new()
+        };
+
+        // Iterate over the line program rows.
+        let mut rows = program.rows();
+        while let Some((header, row)) = rows.next_row()? {
+            if row.end_sequence() {
+                warn!(
+                    "Possible gap in addresses: {:x} end-sequence",
+                    row.address()
+                );
+                continue;
+            }
+
+            // Determine line/column. DWARF line/column is never 0, so we use that
+            // but other applications may want to display this differently.
+            let line = match row.line() {
+                Some(line) => line.get(),
+                None => 0,
+            };
+            let column = match row.column() {
+                gimli::ColumnType::LeftEdge => 0,
+                gimli::ColumnType::Column(column) => column.get(),
+            };
+
+            let ins_index = (row.address() - text_section_offset) / 8;
```

Comment:
```suggestion
            let ins_index = (row.address() - text_section_offset) >> 3;
```

Reply:
This change does nothing, `/ 8` will compile to `shr rax, 3;`. Could we leave the first option for the sake of readability? IMHO `/ 8` is more obvious `uint64` pointer arithmetic than `>> 3` https://godbolt.org/z/41xvK64qs

Reply:
For now I've turned it into this since clippy was complaining on possible panic / UB on underflow.

```rust
if let Some(ins_index) = row
    .address()
    .checked_sub(text_section_offset)
    .map(|x| x / 8)
{
    if hits[ins_index] {
        self.file_coverage(&comp_dir, dwarf, &unit, row, header)
            .hits
            .insert((line, column));
    }
}
```